### PR TITLE
hotfix(admin-api) fixes an edge-case and returns 405 on invalid methods

### PR DESCRIPTION
**Pull request by thefosk**
### Full changelog

* Fixes an edge-case when requesting the Admin API with an invalid method

### CODE REVIEW

**Comment by Tieske** [comment]
The better solution would be to add a metatable to the `methods` table here; https://github.com/Mashape/kong/blob/8f8b4ef3e38029b422f89440f852dd1beace3ce6/kong/api/init.lua#L99-L112

If a lookup fails in that table, it should return a function that executes a 405.

**Comment by Tieske** [comment]
@thibaultcha @thefosk I updated this PR. It seems that Lapis is adding too much magic under the hood for the metatable approach to work properly, hence falling back on the error string checking as implemented by @thefosk.

With the exception that the existing error mechanism (that didn't catch custom http methods) is now removed, as it doubled.

File: kong/api/init.lua [hunk 1]
```diff
@@ -7,125 +7,142 @@ local api_helpers = require "kong.api.api_helpers"
...
+
 app.handle_error = function(self, err, trace)
-  ngx.log(ngx.ERR, err.."\n"..trace)
-  -- We just logged the error so no need to give it to responses and log it twice
-  return responses.send_HTTP_INTERNAL_SERVER_ERROR()
+
+  if err and err:match("don't know how to respond to") then
+    return responses.send_HTTP_METHOD_NOT_ALLOWED()
+
+  else
```
Comment (thibaultcha) [inline kong/api/init.lua]: no `else` here as the previous branch returns

File: kong/api/init.lua [hunk 2]
```diff
@@ -7,125 +7,142 @@ local api_helpers = require "kong.api.api_helpers"
...
 app.handle_404 = function(self)
   return responses.send_HTTP_NOT_FOUND()
 end
 
+
 app.handle_error = function(self, err, trace)
-  ngx.log(ngx.ERR, err.."\n"..trace)
-  -- We just logged the error so no need to give it to responses and log it twice
-  return responses.send_HTTP_INTERNAL_SERVER_ERROR()
+
```
Comment (thibaultcha) [inline kong/api/init.lua]: style: no blank line here

File: kong/api/init.lua [hunk 3]
```diff
@@ -7,125 +7,142 @@ local api_helpers = require "kong.api.api_helpers"
 local tablex = require "pl.tablex"
 
 local find = string.find
-
 local app = lapis.Application()
-
 local needs_body = tablex.readonly({ PUT = 1, POST = 2, PATCH = 3 })
```
Comment (thibaultcha) [inline kong/api/init.lua]: It's good that we are switching to the new code style in this module, but then let's also respect the 2 lines jump here as well: between the `require` section, the "globals caching section", the "Lapis App creation", and the constants definition for this module.

File: kong/api/init.lua [hunk 4]
```diff
@@ -7,125 +7,142 @@ local api_helpers = require "kong.api.api_helpers"
...
 
+
 local handler_helpers = {
   responses = responses,
   yield_error = app_helpers.yield_error
 }
 
+
 local function attach_routes(routes)
+
```
Comment (thibaultcha) [inline kong/api/init.lua]: style: there should be no such blank line when entering a method

File: kong/api/init.lua [hunk 5]
```diff
@@ -7,125 +7,142 @@ local api_helpers = require "kong.api.api_helpers"
...
 
 local find = string.find
-
 local app = lapis.Application()
-
 local needs_body = tablex.readonly({ PUT = 1, POST = 2, PATCH = 3 })
 
+
 local function parse_params(fn)
+
```
Comment (thibaultcha) [inline kong/api/init.lua]: style: blank line when entering a function is not part of the code style

File: kong/api/init.lua [hunk 6]
```diff
@@ -7,125 +7,142 @@ local api_helpers = require "kong.api.api_helpers"
...
 for _, v in ipairs({"kong", "apis", "consumers", "plugins", "cache", "cluster", "certificates", "snis", "upstreams"}) do
   local routes = require("kong.api.routes."..v)
-  attach_routes(insert_405(routes))
+  attach_routes(routes)
 end
 
+
 -- Loading plugins routes
 if singletons.configuration and singletons.configuration.plugins then
+
```
Comment (thibaultcha) [inline kong/api/init.lua]: style: no such blank lines required when entering a block/function

File: kong/api/init.lua [hunk 7]
```diff
@@ -7,125 +7,142 @@ local api_helpers = require "kong.api.api_helpers"
...
   for k in pairs(singletons.configuration.plugins) do
+
```
Comment (thibaultcha) [inline kong/api/init.lua]: ditto: blank line entering a block

File: kong/api/init.lua [hunk 8]
```diff
@@ -7,125 +7,142 @@ local api_helpers = require "kong.api.api_helpers"
...
     local loaded, mod = utils.load_module_if_exists("kong.plugins."..k..".api")
     if loaded then
       ngx.log(ngx.DEBUG, "Loading API endpoints for plugin: ", k)
-      attach_routes(insert_405(mod))
+      attach_routes(mod)
+
     else
       ngx.log(ngx.DEBUG, "No API endpoints loaded for plugin: ", k)
     end
+
```
Comment (thibaultcha) [inline kong/api/init.lua]: No such blank line before closing a block either, or at least, not enforced.

File: kong/api/init.lua [hunk 9]
```diff
@@ -7,125 +7,142 @@ local api_helpers = require "kong.api.api_helpers"
...
   return responses.send_HTTP_NOT_FOUND()
 end
 
+
 app.handle_error = function(self, err, trace)
-  ngx.log(ngx.ERR, err.."\n"..trace)
-  -- We just logged the error so no need to give it to responses and log it twice
-  return responses.send_HTTP_INTERNAL_SERVER_ERROR()
+
+  if err and err:match("don't know how to respond to") then
```
Comment (thibaultcha) [inline kong/api/init.lua]: `string.match` should probably just perform a plain text search if not patterns are needed here.

**Comment by p0pr0ck5** [comment]
Can we avoid style changes in the context of this PR? We could open a separate PR if we care deeply for such style, but adding here seems to introduce unnecessary complexity, dirties the git history, and reduces atomicity of the commit.

**Comment by thibaultcha** [comment]
Re-worked this patch (new history) to properly introduce those 2 different changes, fix + style.

**Pull request merged**